### PR TITLE
Handle corner case of sad64x64 comparison

### LINKED ISSUE #481: SVT HEVC encoder crashes with the particular HEVC standard test sequence SlideShow
latest code crashes at line 660 of EbPictureOperators_C.c with the below command. asm code has the same issue.

-i SlideShow_1280x720.yuv -w 1280 -h 720 -asm 0 -encMode 0 -intra-period 100 -scd 0 -rc 0 -q 22 -n 100 -b SlideShow_1280x720_SVTMODE0_22_SVT.265

### PR DESCRIPTION
**Pull request by intelmark**
Changed the check for pBestSad64x64 to also
include being equal to the value, as the computed
value might equal pBestSad64x64 without a correct
corresponding motion vector ever being set.

We are thus changing to <= to include this corner
case.

Signed-off-by: Mark Feldman <mark.feldman@intel.com>

Closes #481

### CODE REVIEW

File: Source/Lib/C_DEFAULT/EbComputeSAD_C.c [hunk 1]
```diff
@@ -436,7 +436,7 @@ void GetEightHorizontalSearchPointResults_32x32_64x64(
 
         //64x64
         sad64x64 = sad32x32_0 + sad32x32_1 + sad32x32_2 + sad32x32_3;
-        if (2 * sad64x64 < pBestSad64x64[0]){
+        if (2 * sad64x64 <= pBestSad64x64[0]){
```
Comment (Austin-Hu) [inline Source/Lib/C_DEFAULT/EbComputeSAD_C.c]: Don't forget the same changes for GetEightHorizontalSearchPointResults_32x32_64x64_PU_AVX2_INTRIN with AVX2 and AVX512 implementations.

File: Source/Lib/ASM_AVX2/EbComputeSAD_SadLoopKernel_AVX512.c [hunk 2]
```diff
@@ -2180,7 +2180,7 @@ void GetEightHorizontalSearchPointResults_32x32_64x64_PU_AVX2_INTRIN(
 		bestSad64x64 = temSum;
 		bestMV64x64 = 7 * 4;
 	}
-	if (pBestSad64x64[0] != bestSad64x64) {
+	if (bestSad64x64 <= pBestSad64x64[0]) {
```
Comment (Austin-Hu) [inline Source/Lib/ASM_AVX2/EbComputeSAD_SadLoopKernel_AVX512.c]: And the one in Source/Lib/ASM_AVX2/EbComputeSAD_SadLoopKernel_AVX512.c although it's disabled...

**Comment by tianjunwork** [comment]
Hi @intelmark , could you add the test result with `-asm 0`, `-asm 1` with avx2/avx512 here?
Please also check MD5. Thank you.

**Comment by intelmark** [comment]
Testing shows that with -asm 0 and -asm 1, both with avx2 and avx512 encodings with the latest commits in this PR are binary identical.

File: Source/Lib/ASM_AVX2/EbComputeSAD_Intrinsic_AVX2.c [hunk 3]
```diff
@@ -2139,11 +2139,11 @@ void GetEightHorizontalSearchPointResults_32x32_64x64_PU_AVX2_INTRIN(
         bestSad64x64 = temSum;
         bestMV64x64 = 7 * 4;
     }
-    if (bestSad64x64 <= pBestSad64x64[0]) {
-        pBestSad64x64[0] = bestSad64x64;
-        xMv = _MVXT(mv) + (EB_S16)bestMV64x64;  yMv = _MVYT(mv);
-        pBestMV64x64[0] = ((EB_U16)yMv << 16) | ((EB_U16)xMv);
-    }
+
+    pBestSad64x64[0] = bestSad64x64;
```
Comment (tianjunwork) [inline Source/Lib/ASM_AVX2/EbComputeSAD_Intrinsic_AVX2.c]: why removed the if?

**Comment by intelmark** [comment]
Isn’t needed. See the comments in the commit


On Mar 20, 2020, at 9:37 PM, Jun <notifications@github.com> wrote:

﻿

@tianjunwork commented on this pull request.

________________________________

In Source/Lib/ASM_AVX2/EbComputeSAD_Intrinsic_AVX2.c<https://github.com/OpenVisualCloud/SVT-HEVC/pull/500#discussion_r395960848>:

> @@ -2139,11 +2139,11 @@ void GetEightHorizontalSearchPointResults_32x32_64x64_PU_AVX2_INTRIN(
         bestSad64x64 = temSum;
         bestMV64x64 = 7 * 4;
     }
-    if (bestSad64x64 <= pBestSad64x64[0]) {
-        pBestSad64x64[0] = bestSad64x64;
-        xMv = _MVXT(mv) + (EB_S16)bestMV64x64;  yMv = _MVYT(mv);
-        pBestMV64x64[0] = ((EB_U16)yMv << 16) | ((EB_U16)xMv);
-    }
+
+    pBestSad64x64[0] = bestSad64x64;


why removed the if?

**Comment by tianjunwork** [comment]
Hi @tszumski , could you help review the PR? Thank you.

**Comment by tianjunwork** [comment]
Hi @intelmark , please rerun the CI fast type. And also check for perf. Thank you.

**Comment by tszumski** [comment]
Hi Mark,
Changes looks good but I have different proposition for AVX changes (less code):
https://github.com/tszumski/SVT-HEVC/commit/80362a5b909dcce6577145b98881ecbd02118177

File: Source/Lib/ASM_AVX2/EbComputeSAD_SadLoopKernel_AVX512.c [hunk 4]
```diff
@@ -2136,55 +2136,56 @@ void GetEightHorizontalSearchPointResults_32x32_64x64_PU_AVX2_INTRIN(
...
-	bestSad64x64 = pBestSad64x64[0];
-	bestMV64x64 = 0;
+	bestMV64x64 = 0xff;
+
 	//sad_0
 	temSum = _mm_extract_epi32(sad_0, 0);
-	if (temSum < bestSad64x64){
-		bestSad64x64 = temSum;
+	if (temSum <= pBestSad64x64[0]){
+		pBestSad64x64[0] = temSum;
```
Comment (tszumski) [inline Source/Lib/ASM_AVX2/EbComputeSAD_SadLoopKernel_AVX512.c]: bestMV64x64 = 0;

**Comment by intelmark** [comment]
python SVT-HEVC_FunctionalTests.py Fast -type asm_test

Test Begin...
Total Number of Tests: 108
Total Passed: 108
Percentage Passed: 100.0%
Time Elapsed: 9 minute(s), 53 second(s) 

Speed check:
(with change)
Bin/Release/SvtHevcEncApp -i /testfiles/bbb_1920x1080_420p.yuv -w 1920 -h 1080 -asm 0 -encMode 0 -intra-period 100 -scd 0 -rc 0 -q 22 -n 100 -b Slibbb.265
Average Speed:          2.11 fps
Total Encoding Time:    47446 ms
Total Execution Time:   47952 ms
Average Latency:        33999 ms
Max Latency:            46614 ms

(master branch)
Average Speed:          2.08 fps
Total Encoding Time:    48028 ms
Total Execution Time:   48529 ms
Average Latency:        34168 ms
Max Latency:            47832 ms
